imap/monitor.go

```diff
@@ -147,48 +151,91 @@ func checkForNewEmails(im models.IMAP) {
-					reportingFailed = append(reportingFailed, m.UID)
-				} else {
-					err = result.HandleEmailReport(models.EventDetails{})
-					if err != nil {
-						log.Error("Error updating GoPhish email with rid ", rid, ": ", err.Error())
-					} else {
-						if im.DeleteReportedCampaignEmail == true {
-							campaignEmails = append(campaignEmails, m.UID)
-						}
-					}
+					reportingFailed = append(reportingFailed, m.SeqNum)
+					continue
 				}
-			} else {
-				// In the future this should be an alert in Gophish
-				log.Debugf("User '%s' reported email with subject '%s'. This is not a GoPhish campaign; you should investigate it.\n", m.Email.From, m.Email.Subject)
-			}
-			// Check if any emails were unable to be reported, so we can mark them as unread
-			if len(reportingFailed) > 0 {
-				log.Debugf("Marking %d emails as unread as failed to report\n", len(reportingFailed))
-				err := mailServer.MarkAsUnread(reportingFailed) // Set emails as unread that we failed to report to GoPhish
+				err = result.HandleEmailReport(models.EventDetails{})
 				if err != nil {
-					log.Error("Unable to mark emails as unread: ", err.Error())
+					log.Error("Error updating GoPhish email with rid ", rid, ": ", err.Error())
+					continue
 				}
-			}
-			// If the DeleteReportedCampaignEmail flag is set, delete reported Gophish campaign emails
-			if im.DeleteReportedCampaignEmail == true && len(campaignEmails) > 0 {
-				log.Debugf("Deleting %d campaign emails\n", len(campaignEmails))
-				err := mailServer.DeleteEmails(campaignEmails) // Delete GoPhish campaign emails.
-				if err != nil {
-					log.Error("Failed to delete emails: ", err.Error())
+				if im.DeleteReportedCampaignEmail == true {
+					deleteEmails = append(deleteEmails, m.SeqNum)
 				}
 			}
+			
+		}
+		// Check if any emails were unable to be reported, so we can mark them as unread
+		if len(reportingFailed) > 0 {
+			log.Debugf("Marking %d emails as unread as failed to report", len(reportingFailed))
+			err := mailServer.MarkAsUnread(reportingFailed) // Set emails as unread that we failed to report to GoPhish
+			if err != nil {
+				log.Error("Unable to mark emails as unread: ", err.Error())
+			}
 		}
+		// If the DeleteReportedCampaignEmail flag is set, delete reported Gophish campaign emails
+		if len(deleteEmails) > 0 {
+			log.Debugf("Deleting %d campaign emails", len(deleteEmails))
+			err := mailServer.DeleteEmails(deleteEmails) // Delete GoPhish campaign emails.
+			if err != nil {
+				log.Error("Failed to delete emails: ", err.Error())
+			}
+		}
+		
 	} else {
 		log.Debug("No new emails for ", im.Username)
 	}
 }
+
+func checkRIDs(em *email.Email, rids map[string]int){
+
+	// Check Text and HTML
+	emailContent := string(em.Text) + string(em.HTML)
+	for _, r := range goPhishRegex.FindAllStringSubmatch(emailContent, -1) {
+		newrid := r[len(r)-1]
+		if _, ok := rids[newrid]; !ok {
+			rids[newrid] = 0
+		}
+		rids[newrid]++
```

Comment:
So I thought a bit more about this, and this is really just being used as a set. To that end, we don't actually need to keep track of the number of times we've seen this rid.

This let's us make things even easier, since the ++ line isn't need. To make things _even simpler_ we can use the fact that in Go, retrieving a value using a key that doesn't exist gives you the zero value for that value's type. You can see an example [here](https://blog.golang.org/maps#TOC_4.).

This means that we can just change `rids` to be a `map[string]bool` and then do something like:

```go
if !rids[newrid] {
    rids[newrid] = true
}
```

which is much easier to read.

As a side note, some people may prefer to use `map[string]struct{}` and do the normal `!ok` check since empty `struct{}`'s take up 0 bytes, but personally I like the readability of using a bool, and we're not storing a lot of items here so the cost is negligible, IMO.

Reply:
I like this approach a lot, done!